### PR TITLE
Add elt pupil

### PR DESCRIPTION
**Pull request by syhaffert**
I have added a model of the ELT pupil based on the ELT contract proposal ( https://www.eso.org/sci/facilities/eelt/docs/index.html ). And I also added a paraxial thin lens function.

### CODE REVIEW

**Comment by syhaffert** [comment]
Okay, I forgot what I added in this branch. I think I will need to add some more documentation before we can merge.

**Comment by syhaffert** [comment]
Everything has now been updated. This pull request will have a major impact on the Pyramid wavefront sensor. I have changed the interface to make it the same as the ODWFS interface. The proposed interface is more flexible and easier to use, especially when you want to have certain symmetries in the PWFS response.

File: hcipy/aperture/realistic.py
Comment (ehpor) [inline hcipy/aperture/realistic.py:586]: Make 39.14634 a variable instead of copy-pasting it all over the place.

File: hcipy/aperture/realistic.py
Comment (ehpor) [inline hcipy/aperture/realistic.py:590]: pi/2 * 0?

File: hcipy/atmosphere/standard_atmosphere.py
Comment (ehpor) [inline hcipy/atmosphere/standard_atmosphere.py:17]: No acronyms in function names. Maybe loose the "observatory" if you think it's too long?

File: hcipy/atmosphere/standard_atmosphere.py
Comment (ehpor) [inline hcipy/atmosphere/standard_atmosphere.py:47]: No empty line here.

File: hcipy/atmosphere/standard_atmosphere.py
Comment (ehpor) [inline hcipy/atmosphere/standard_atmosphere.py:58]: Always end a file with a new line.

File: hcipy/data/eelt_segment_indices.txt
Comment (ehpor) [inline hcipy/data/eelt_segment_indices.txt:1]: Is there any way to do this programmatically? I'd imagine you didn't type all this by hand.

File: hcipy/field/util.py
Comment (ehpor) [inline hcipy/field/util.py:363]: Why add a nanmedian as well? Also, modify the documentation for this if you want to keep it.

File: hcipy/metrics/profile.py
Comment (ehpor) [inline hcipy/metrics/profile.py:57]: Correct the function docstring as well to reflect this change.

File: hcipy/mode_basis/LP_fiber_modes.py
Comment (ehpor) [inline hcipy/mode_basis/LP_fiber_modes.py:180]: Why? There is an if-else already for returns.

File: hcipy/optics/deformable_mirror.py
Comment (ehpor) [inline hcipy/optics/deformable_mirror.py:158]: No empty line here.

File: hcipy/optics/deformable_mirror.py
Comment (ehpor) [inline hcipy/optics/deformable_mirror.py:165]: Only one empty line here.

File: hcipy/optics/optics.py
Comment (ehpor) [inline hcipy/optics/optics.py:22]: No new line here.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:31]: Only one empty line between functions.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:62]: Only one empty line between functions.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:97]: End a file on a new line.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:95]: No new line here.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:80]: Variable names should be lower case. You should use `num_x` here, I think.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:82]: Formatting.

File: hcipy/wavefront_sensing/pyramid.py
Comment (ehpor) [inline hcipy/wavefront_sensing/pyramid.py:69]: Rename this to `pupil_diameter`.

File: hcipy/wavefront_sensing/pyramid.py
Comment (ehpor) [inline hcipy/wavefront_sensing/pyramid.py:96]: The extent of a regular grid is not `x.ptp()` but `grid.delta * grid.dims`. What about elongated pupils?

File: hcipy/wavefront_sensing/pyramid.py
Comment (ehpor) [inline hcipy/wavefront_sensing/pyramid.py:208]: Formatting.

File: tests/test_aperture.py
Comment (ehpor) [inline tests/test_aperture.py:144]: End a file on a new line.

File: hcipy/aperture/realistic.py
Comment (ehpor) [inline hcipy/aperture/realistic.py:573]: Should be lower case.

File: hcipy/aperture/realistic.py
Comment (ehpor) [inline hcipy/aperture/realistic.py:598]: Formatting.

File: hcipy/aperture/realistic.py
Comment (ehpor) [inline hcipy/aperture/realistic.py:600]: Formatting.

File: hcipy/aperture/realistic.py
Comment (ehpor) [inline hcipy/aperture/realistic.py:597]: The 0.4 should probably be a `spider_width` variable.

File: hcipy/aperture/realistic.py
Comment (ehpor) [inline hcipy/aperture/realistic.py:627]: End a file on a new line.

File: hcipy/optics/optics.py
Comment (ehpor) [inline hcipy/optics/optics.py:1]: Any better name for this file?

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:48]: Variable names should be lower case. You should use num_x here, I think.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:24]: Variable names should be lower case. You should use num_x here, I think.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:23]: No new line here. Also, check the properties of your grid.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:28]: Formatting.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:59]: No new line here.

**Review by ehpor — CHANGES_REQUESTED** [review]
I think that's most of it. Also, look at codacy (https://app.codacy.com/gh/ehpor/hcipy/pullRequest?bid=22285194&prid=6970167) to fix the new issues that popped up there.

**Comment by ehpor** [comment]
Oh, and for the future, please split up your PR into multiple small independent PRs. This one does way too many separate things.

File: hcipy/data/eelt_segment_indices.txt
Comment (syhaffert) [reply hcipy/data/eelt_segment_indices.txt:1]: Yes, I did it with a small function. But I do not have any guarantees at the moment that this function always creates the correct mask. I tweaked the parameters until the aperture matched the documentation. Should we keep it like this for now? I would first need to verify that the function works for any combination of parameters.

File: hcipy/field/util.py
Comment (syhaffert) [reply hcipy/field/util.py:363]: I use hcipy for some data reduction pipelines and I needed the nanmedian for something.

File: hcipy/data/eelt_segment_indices.txt
Comment (ehpor) [reply hcipy/data/eelt_segment_indices.txt:1]: What do you mean with "any combination of parameters"? There is only one ELT pupil that it needs to generate, right?

File: hcipy/aperture/realistic.py
Comment (syhaffert) [reply hcipy/aperture/realistic.py:600]: I have changed the diameter to a variable. Do you mean more formatting? a space to change [0,0] -> [0, 0]?

File: hcipy/field/util.py
Comment (ehpor) [reply hcipy/field/util.py:363]: Okay, then add in the documentation that this ignores NaNs inside the superpixel.

File: hcipy/optics/optics.py
Comment (syhaffert) [reply hcipy/optics/optics.py:1]: optical_components ?

File: hcipy/aperture/realistic.py
Comment (ehpor) [reply hcipy/aperture/realistic.py:600]: Yes, use PEP8 formatting, everywhere and always.

File: hcipy/optics/optics.py
Comment (ehpor) [reply hcipy/optics/optics.py:1]: `thin_lens.py`?

File: hcipy/optics/optics.py
Comment (ehpor) [inline hcipy/optics/optics.py:5]: Maybe make this a `ThinLens` class?

File: hcipy/data/eelt_segment_indices.txt
Comment (syhaffert) [reply hcipy/data/eelt_segment_indices.txt:1]: Yes, but my segment finding algorithm finds the masks based on a relative illumination pattern. And this depends on the sampling density and the chosen power cut off. It will fail for  small number of samples. Because I have the actual pupil now, I can probably create a better subset criterion.

File: hcipy/optics/optics.py
Comment (syhaffert) [reply hcipy/optics/optics.py:5]: Done

File: hcipy/data/eelt_segment_indices.txt
Comment (syhaffert) [reply hcipy/data/eelt_segment_indices.txt:1]: Remove the file dependency. It is now based on programmed rules.

File: hcipy/aperture/realistic.py
Comment (ehpor) [inline hcipy/aperture/realistic.py:110]: I did not see this the first time. `oversize`?

File: hcipy/aperture/realistic.py
Comment (ehpor) [reply hcipy/aperture/realistic.py:753]: You marked as resolved, but I see no newline here.

File: hcipy/atmosphere/standard_atmosphere.py
Comment (ehpor) [reply hcipy/atmosphere/standard_atmosphere.py:57]: You marked as resolved, but I see no newline here.

File: hcipy/field/util.py
Comment (ehpor) [reply hcipy/field/util.py:363]: You marked as resolved, but I see no mention of ignoring nans in the documentation on this line.

File: hcipy/mode_basis/LP_fiber_modes.py
Comment (ehpor) [inline hcipy/mode_basis/LP_fiber_modes.py:178]: End this file on a newline.

File: hcipy/util/finite_difference.py
Comment (ehpor) [reply hcipy/util/finite_difference.py:91]: You marked this as resolved, but I see no newline here.

File: tests/test_aperture.py
Comment (ehpor) [reply tests/test_aperture.py:187]: You marked this as resolved, but I see no newline here.

File: tests/test_optics.py
Comment (ehpor) [inline tests/test_optics.py:678]: End the file on a newline.

**Review by ehpor — CHANGES_REQUESTED** [review]
Only some extremely small style changes remaining (plus one documentation and one straggling function argument.

Files have to end on a blank line to follow PEP8.

**Comment by ehpor** [comment]
@syhaffert Oh, and while your doing this: please also fix the new code quality issues as listed on https://app.codacy.com/gh/ehpor/hcipy/pullRequest?prid=6970167. They should have showed up in the automated checks, but for some reason don't. I'll have to check that out.

File: hcipy/aperture/realistic.py
Comment (syhaffert) [reply hcipy/aperture/realistic.py:110]: That has to be removed. Done

**Comment by syhaffert** [comment]
I fixed the issues that Codacy reported, except for 3 trailing white spaces that I could not find.

**Review by ehpor — DISMISSED** [review]

**Comment by ehpor** [comment]
That's fine. I just don't want to accumulate these at an alarming rate every PR. I'll merge after the PR check finishes on Azure. Don't know why it's not showing up as a required check; it's supposed to, but only the commit CI checks are listed for some reason.

**Comment by syhaffert** [comment]
I have written a couple more tests to increase the test coverage because this PR failed that check.

File: hcipy/atmosphere/standard_atmosphere.py
Comment (syhaffert) [reply hcipy/atmosphere/standard_atmosphere.py]: There is a new line now.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:1]: Merge this with the line below.

File: tests/test_optics.py
Comment (ehpor) [inline tests/test_optics.py:646]: How did this pass the test before?

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:23]: You are now relying on `kernel` being a `Field`. Change the documentation to reflect this.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:26]: Why use `np.meshgrid()` when you can get the same values from `kernel.grid`?

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:28]: Any check on whether `kernel.grid` has the same sample spacing (ie. `grid.delta`) as `grid`?

**Review by ehpor — CHANGES_REQUESTED** [review]

File: tests/test_optics.py
Comment (syhaffert) [reply tests/test_optics.py:646]: No clue? Are the tests in this file included in the unit tests?

Comment (ehpor) [reply tests/test_optics.py:646]: I think so. Anything in the `/tests` directory is auto-enumerated, as long as the function starts with `test_`. Let me check if it ran on the old CI.

Comment (ehpor) [reply tests/test_optics.py:646]: Oh, it didn't pass. The CI tests failed for the commit before this one, with an error in this function. So all is fine.

Comment (syhaffert) [reply tests/test_optics.py:646]: I found it. In commit #67b6ab9 I added more tests, and that commit did not pass the unit tests. So a commit later I fixed the bugs.

File: hcipy/util/finite_difference.py
Comment (syhaffert) [reply hcipy/util/finite_difference.py:1]: Done.

Comment (ehpor) [reply hcipy/util/finite_difference.py:1]: I'll click resolve after I confirm you made the change. Don't resolve yourself.

File: hcipy/util/finite_difference.py
Comment (syhaffert) [reply hcipy/util/finite_difference.py:26]: kernel.grid returns the x and y values in physical units. I need to have the values in index units to make sure the kernel is added in the correct places in the sparse convolution matrix.

File: hcipy/util/finite_difference.py
Comment (syhaffert) [reply hcipy/util/finite_difference.py:23]: I have added a hasattr to check if it is a field.

File: hcipy/util/finite_difference.py
Comment (syhaffert) [reply hcipy/util/finite_difference.py:28]: Done.

File: hcipy/util/finite_difference.py
Comment (ehpor) [reply hcipy/util/finite_difference.py:23]: No need to explicitly check. It's just that the documentation still says `kernel` needs to be `array_like`, while it actually needs to be a `Field`.

File: hcipy/util/finite_difference.py
Comment (ehpor) [inline hcipy/util/finite_difference.py:23]: Use `if not np.allclose(kernel.grid.delta, grid.delta)`. Its intentions are clearer and it's a little more advanced.

File: hcipy/util/finite_difference.py
Comment (ehpor) [reply hcipy/util/finite_difference.py:36]: `np.indices()` then?

File: hcipy/util/finite_difference.py
Comment (ehpor) [reply hcipy/util/finite_difference.py]: Documentation is unchanged... `kernel` can be both a Field and a ndarray, but the documentation still says only ndarray.

File: hcipy/util/finite_difference.py
Comment (syhaffert) [reply hcipy/util/finite_difference.py]: yeah that is a nicer function. I also changed the variables names to better reflect what they are.

File: hcipy/util/finite_difference.py
Comment (syhaffert) [reply hcipy/util/finite_difference.py]: Done

File: hcipy/util/finite_difference.py
Comment (syhaffert) [reply hcipy/util/finite_difference.py]: Changed the documentation to "Field or array_like",

**Review by ehpor — APPROVED** [review]

**Comment by ehpor** [comment]
LGTM. Waiting for CI, but that should pass just fine.

**Pull request merged**
